### PR TITLE
[Refactor]: Core/API Split

### PR DESCRIPTION
**Pull request by dave-gray101**
I've ran into circular reference problems a few times while attempting larger projects within LocalAI - so I decided it was time for a major refactor to try and flatten out our dependency hierarchy into a more orderly chain. This way, we can more easily add functionality to the system!

In addition to the refactoring to change up the references, I also refactored the codebase to have a strict split between the HTTP/endpoint code and the actual LocalAI functionality. 

There is also a _bare stub_ of the MQTT functionality at this point - while it looks useless, it's actually quite important - it is there to help test for circular references while I refactor!

Below I've pasted a copy of the directory structure info I've added to the devdocs, since it also explains the refactoring here.

### Directory Structure of this Repo

The core repository is broken up into the following primary chunks:

* `/backend`: gRPC protobuf specification and gRPC backends. Subfolders for each language.
* **`/core`**: golang sourcecode for the core LocalAI application. Broken down below.
* `/docs`: localai.io website that you are reading now
* `/examples`: example code integrating LocalAI to other projects and/or developer samples and tools
* `/internal`: **here be dragons**. Don't touch this, it's used for automatic versioning.
* `/models`: _No code here!_ This is where models are installed!
* **`/pkg`**: golang sourcecode that is intended to be reusable or at least widely imported across LocalAI. Broken down below
* `/prompt-templates`: _No code here!_ This is where **example** prompt templates were historically stored. Somewhat obsolete these days, model-galleries tend to replace manually creating these?
* `/tests`: Does what it says on the tin. Please write tests and put them here when you do.

The `core` folder is broken down further:

* **`/core/backend`**: code that interacts with a gRPC backend to perform AI tasks.
* `/core/http`: code specifically related to the REST server
* `/core/http/endpoints`: Has two subdirectories, `openai` and `localai` for binding the respective endpoints to the correct backend or service.
* `/core/mqtt`: core specifically related to the MQTT server. Stub for now. Coming soon!
* **`/core/services`**: code implementing functionality performed by `local-ai` itself, rather than delegated to a backend.
* `/core/startup`: code related specifically to application startup of `local-ai`. Potentially to be refactored to become a part of `/core/services` at a later date, or not.

The `pkg` folder is broken down further:

* `/pkg/assets`: Currently contains a single function related to extracting files from archives. Potentially to be refactored to become a part of `/core/utils` at a later date?
* **`/pkg/datamodel`**: Contains the data types and definitions used by the LocalAI project. Imported widely!
* `/pkg/gallery`: Code related to interacting with a `model-gallery`
* `/pkg/grammar`: Code related to BNF / functions for LLM
* `/pkg/grpc`: base classes and interfaces for gRPC backends to implement
* `/pkg/langchain`: langchain related code in golang
* **`/pkg/model`**: Code related to loading and initializing a model and creating the appropriate gRPC backend.
* `/pkg/stablediffusion`: Code related to stablediffusion in golang.
* `/pkg/utils`: Every real programmer knows what they are going to find in here... it's our junk drawer of utility functions.

### CODE REVIEW

**Comment by mudler** [comment]
please stand-by @dave-gray101 the PR is quite big so will take some time for me to review it, but will try to have a look at it asap to avoid merge conflicts

**Comment by mudler** [comment]
@dave-gray101 had to cut 2.3.1 with a couple of bugfixes so we can safely merge this too. Although now there are conflicts (sorry!)

**Comment by dave-gray101** [comment]
@mudler Pr has been updated, merged, and passes CI if now is a more clear merge window. Thanks!

File: docs/content/advanced/development.md [hunk 1]
```diff
@@ -17,6 +17,53 @@ This section will collect how-to, notes and development documentation
...
+## LocalAI Project Structure
+
+**LocalAI is made of multiple components, developed in multiple repositories:**
+
+The core repository, containing the primary `local-ai` server code, gRPC stubs, this documentation website, and docker container building resources are all located at [mudler/LocalAI](https://github.com/mudler/LocalAI).
+
+As LocalAI is designed to make use of multiple, independent model galleries, those are maintained seperately. The following public model galleries are available for use:
+
+* [go-skynet/model-gallery](https://github.com/go-skynet/model-gallery) - The original gallery, the `golang` huggingface scraper ran into limits and was largely retired, so this now holds handmade yaml configs
+* [dave-gray101/model-gallery](https://github.com/dave-gray101/model-gallery) - An automated gallery designed to track HuggingFace uploads and produce best-effort automatically generated configurations for LocalAI. It is designed to produce one LocalAI gallery per repository on HuggingFace.
```
Comment (mudler) [inline docs/content/advanced/development.md:29]: that's a great addition too, I don't want that to go down unnoticed, maybe also add it in the README? https://github.com/mudler/LocalAI#-community-and-integrations

File: core/backend/image.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	log.Debug().Msgf("Parameter Config: %+v", config)
+
+	switch config.Backend {
+	case "stablediffusion":
+		config.Backend = model.StableDiffusionBackend
+	case "tinydream":
+		config.Backend = model.TinyDreamBackend
+	case "":
+		config.Backend = model.StableDiffusionBackend
+	default:
```
Comment (mudler) [inline core/backend/image.go:114]: this is actually a no-op, better drop  default entirely?

Comment (mudler) [reply core/backend/image.go:114]: We had also issues around it, would actually avoid to re-introduce it https://github.com/mudler/LocalAI/pull/1500

File: pkg/datamodel/openai.go [hunk 3]
```diff
@@ -1,10 +1,8 @@
-package schema
+package datamodel
```
Comment (mudler) [inline pkg/datamodel/openai.go:1]: I think that we should keep `schema`, or even go with `modelrequest`. I think `datamodel` is too much generic here (at beginning it almost confused me here)

File: pkg/gallery/models.go [hunk 4]
```diff
@@ -183,54 +179,3 @@ func InstallModel(basePath, nameOverride string, config *Config, configOverrides
 
 	return nil
 }
-
-type progressWriter struct {
```
Comment (mudler) [inline pkg/gallery/models.go:187]: trying to understand where you moved the `progressWriter`, the diff is quite huge for Github hand hard to follow up

**Comment by mudler** [comment]
ok I give it a first-pass, but even if the changes are refactors/moving functions around, github makes quite difficult to read the diff, hang on and will check it once more, but it's looking good. I'll have a look later at it again

File: core/backend/image.go (hunk 2 above)
Comment (mudler) [reply core/backend/image.go:114]: ah yes - this is actually an issue. If we set a default here we leave the external backends in the cold.

Comment (mudler) [reply core/backend/image.go:114]:
```suggestion
```

File: pkg/datamodel/openai.go (hunk 3 above)
Comment (dave-gray101) [reply pkg/datamodel/openai.go:1]: I went with a super generic name admittedly - I'm OK with pkg/schema instead!

I assume this applies to the whole pkg/datamodel, not just openai.go, and will refactor it as such :)

File: pkg/gallery/models.go (hunk 4 above)
Comment (dave-gray101) [reply pkg/gallery/models.go:187]: processWriter got moved down into pkg/utils/uri.go since unless I made a merging error somewhere along the way, it's only used within uri.go's `DownloadFile` function

**Review by mudler — APPROVED** [review]

**Comment by mudler** [comment]
@dave-gray101 I think it's missing https://github.com/mudler/LocalAI/pull/1526

**Comment by mudler** [comment]
re-adding it back in #1532

**Comment by mudler** [comment]
A closer look made me realize this changeset skipped entirely tests as moving things to `core` would not trigger test suites: https://github.com/mudler/LocalAI/blob/ab7b4d5ee9448e533a342bd1771393acd2967191/Makefile#L339

I'm trying to re-add it in #1532, but if many failures are getting in the way we might need to revert it

**Comment by mudler** [comment]
also parts of: https://github.com/mudler/LocalAI/commit/c1888a8062fc7a000457714d8426584d42d0f142
